File: c++/0486.cpp
// Link: https://leetcode.com/problems/predict-the-winner/description/
// Runtime: 3 ms
// Memory: 7.4 MB

class Solution 
{
public:
    bool PredictTheWinner(const std::vector<int>& nums) 
    {
        std::vector<int> scores = nums;
        for (int difference = 1; difference < static_cast<int>(nums.size()); difference++)
            for (int left = 0; left < static_cast<int>(nums.size() - difference); left++)
                scores[left] = std::max(nums[left] - scores[left + 1], nums[left + difference] - scores[left]);

        return scores[0] >= 0;
    }
};

File: python/0486.py
# Link: https://leetcode.com/problems/minimum-size-subarray-sum/description/
# Runtime: 34 ms
# Memory: 16.27 MB

class Solution:
    def PredictTheWinner(self, nums: List[int]) -> bool:
        scores = nums[:]
        for difference in range(1, len(nums)):
            for left in range(len(nums) - difference):
                scores[left] = max(nums[left] - scores[left + 1], nums[left + difference] - scores[left])
        
        return scores[0] >= 0
